File: tools/pkg/acltester/rp.go
package acltester

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	corev1 "k8s.io/api/core/v1"
)

type RPClient struct {
	HTTPClient   *http.Client
	APIServerURL string
	NodeName     string
}

func (c *RPClient) RunScenario(ctx context.Context, scenario *Scenario) (*ScenarioReport, error) {
	report := &ScenarioReport{Steps: make([]StepReport, 0, len(scenario.Steps)), Passed: true}
	for index, step := range scenario.Steps {
		started := time.Now()
		stepReport, err := c.runStep(ctx, index, step)
		if err != nil {
			return nil, err
		}
		stepReport.ElapsedMS = time.Since(started).Milliseconds()
		report.Steps = append(report.Steps, *stepReport)
		report.Passed = report.Passed && stepReport.Passed
	}
	return report, nil
}

func (c *RPClient) runStep(ctx context.Context, index int, step ScenarioStep) (*StepReport, error) {
	switch {
	case step.Patch != nil:
		if err := c.patchNodeLabels(ctx, step.Patch.Labels()); err != nil {
			return nil, err
		}
		return &StepReport{Index: index, Kind: "patch", Passed: true, Message: "patched fake Node labels"}, nil
	case step.Expect != nil:
		return c.expectState(ctx, index, step.Expect)
	case step.AssertFailureReason != "":
		node, err := c.getNode(ctx)
		if err != nil {
			return nil, err
		}
		actual := node.Labels[FailureReasonLabel]
		passed := actual == step.AssertFailureReason
		message := "failure reason matched"
		if !passed {
			message = "failure reason mismatch"
		}
		return &StepReport{
			Index:    index,
			Kind:     "assert-failure-reason",
			Passed:   passed,
			Message:  message,
			Expected: step.AssertFailureReason,
			Actual:   actual,
		}, nil
	default:
		return nil, fmt.Errorf("step %d had no recognized action", index)
	}
}

func (c *RPClient) expectState(ctx context.Context, index int, step *ExpectStep) (*StepReport, error) {
	deadline := time.Now().Add(step.Timeout)
	pollInterval := 500 * time.Millisecond
	var lastObserved map[string]string
	matched := false

	for time.Now().Before(deadline) {
		node, err := c.getNode(ctx)
		if err != nil {
			return nil, err
		}
		lastObserved = map[string]string{
			"state":               node.Labels[StateLabel],
			"observed-request-id": node.Labels[ObservedRequestIDLabel],
			"failure-reason":      node.Labels[FailureReasonLabel],
		}
		if node.Labels[StateLabel] == step.State {
			if step.ObservedRequestID == "" || node.Labels[ObservedRequestIDLabel] == step.ObservedRequestID {
				matched = true
				break
			}
		}
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(pollInterval):
		}
	}

	passed := matched
	message := "observed expected state"
	if step.ExpectTimeout {
		passed = !matched
		message = "timed out as expected"
	}
	if !passed && !step.ExpectTimeout {
		message = "state expectation failed"
	}
	return &StepReport{
		Index:   index,
		Kind:    "expect",
		Passed:  passed,
		Message: message,
		Expected: map[string]any{
			"state":               step.State,
			"observed-request-id": step.ObservedRequestID,
			"timeout":             step.Timeout.String(),
			"expect-timeout":      step.ExpectTimeout,
		},
		Actual: lastObserved,
	}, nil
}

func (c *RPClient) patchNodeLabels(ctx context.Context, labels map[string]string) error {
	body, err := json.Marshal(map[string]any{
		"metadata": map[string]any{
			"labels": labels,
		},
	})
	if err != nil {
		return err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPatch, c.nodeURL(), bytes.NewReader(body))
	if err != nil {
		return err
	}
	request.Header.Set("Content-Type", "application/merge-patch+json")
	response, err := c.client().Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	if response.StatusCode >= 300 {
		return fmt.Errorf("fake apiserver patch failed with %s", response.Status)
	}
	return nil
}

func (c *RPClient) getNode(ctx context.Context) (*corev1.Node, error) {
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, c.nodeURL(), nil)
	if err != nil {
		return nil, err
	}
	response, err := c.client().Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()
	if response.StatusCode >= 300 {
		return nil, fmt.Errorf("fake apiserver get failed with %s", response.Status)
	}
	var node corev1.Node
	if err := json.NewDecoder(response.Body).Decode(&node); err != nil {
		return nil, fmt.Errorf("failed to decode fake Node response: %w", err)
	}
	return &node, nil
}

func (c *RPClient) nodeURL() string {
	return strings.TrimRight(c.APIServerURL, "/") + "/api/v1/nodes/" + c.NodeName
}

func (c *RPClient) client() *http.Client {
	if c.HTTPClient != nil {
		return c.HTTPClient
	}
	return http.DefaultClient
}

File: crates/trident-acl-agent-tester/src/rp.rs
//! RP-side scenario runner for the Harpoon label protocol.

use std::{collections::BTreeMap, time::Instant};

use anyhow::Context;
use reqwest::Client;
use serde_json::json;
use tokio::time::{sleep, Duration, Instant as TokioInstant};
use url::Url;

use crate::scenario::{labels_from_patch_step, Scenario, ScenarioReport, ScenarioStep, StepReport};

const STATE_LABEL: &str = "kubernetes.azure.com/trident-abupdate-state";
const OBSERVED_REQUEST_ID_LABEL: &str = "kubernetes.azure.com/trident-abupdate-observed-request-id";
const FAILURE_REASON_LABEL: &str = "kubernetes.azure.com/trident-abupdate-failure-reason";

pub async fn run_scenario(
    client: &Client,
    apiserver_url: &Url,
    node_name: &str,
    scenario: &Scenario,
) -> anyhow::Result<ScenarioReport> {
    let mut steps = Vec::with_capacity(scenario.steps.len());
    let mut passed = true;

    for (index, step) in scenario.steps.iter().enumerate() {
        let start = Instant::now();
        let report = match step {
            ScenarioStep::Patch { patch } => {
                let labels = labels_from_patch_step(patch);
                patch_node_labels(client, apiserver_url, node_name, labels).await?;
                StepReport {
                    index,
                    kind: "patch".to_string(),
                    passed: true,
                    elapsed_ms: start.elapsed().as_millis(),
                    message: "patched fake Node labels".to_string(),
                    expected: None,
                    actual: None,
                }
            }
            ScenarioStep::Expect { expect } => {
                let deadline = TokioInstant::now() + expect.timeout.0;
                let expected = json!({
                    "state": expect.state,
                    "observedRequestId": expect.observed_request_id,
                    "expectTimeout": expect.expect_timeout,
                });
                let mut last_seen = json!({});
                let mut matched = false;

                while TokioInstant::now() < deadline {
                    let node = get_node(client, apiserver_url, node_name).await?;
                    let labels = label_map(&node);
                    last_seen = json!({
                        "state": labels.get(STATE_LABEL),
                        "observedRequestId": labels.get(OBSERVED_REQUEST_ID_LABEL),
                        "failureReason": labels.get(FAILURE_REASON_LABEL),
                    });
                    if labels.get(STATE_LABEL).map(String::as_str) == Some(expect.state.as_str())
                        && expect
                            .observed_request_id
                            .as_deref()
                            .map(|expected| {
                                labels.get(OBSERVED_REQUEST_ID_LABEL).map(String::as_str)
                                    == Some(expected)
                            })
                            .unwrap_or(true)
                    {
                        matched = true;
                        break;
                    }
                    sleep(Duration::from_millis(500)).await;
                }

                let passed_step = if expect.expect_timeout {
                    !matched
                } else {
                    matched
                };
                StepReport {
                    index,
                    kind: "expect".to_string(),
                    passed: passed_step,
                    elapsed_ms: start.elapsed().as_millis(),
                    message: if passed_step {
                        if expect.expect_timeout {
                            "timed out as expected".to_string()
                        } else {
                            "observed expected state".to_string()
                        }
                    } else {
                        "state expectation failed".to_string()
                    },
                    expected: Some(expected),
                    actual: Some(last_seen),
                }
            }
            ScenarioStep::AssertFailureReason {
                assert_failure_reason,
            } => {
                let node = get_node(client, apiserver_url, node_name).await?;
                let labels = label_map(&node);
                let actual = labels.get(FAILURE_REASON_LABEL).map(|v| v.to_string());
                let passed_step = actual.as_deref() == Some(assert_failure_reason.as_str());
                StepReport {
                    index,
                    kind: "assert-failure-reason".to_string(),
                    passed: passed_step,
                    elapsed_ms: start.elapsed().as_millis(),
                    message: if passed_step {
                        "failure reason matched".to_string()
                    } else {
                        "failure reason mismatch".to_string()
                    },
                    expected: Some(json!(assert_failure_reason)),
                    actual: Some(json!(actual)),
                }
            }
        };

        passed &= report.passed;
        steps.push(report);
    }

    Ok(ScenarioReport { passed, steps })
}

async fn patch_node_labels(
    client: &Client,
    apiserver_url: &Url,
    node_name: &str,
    labels: BTreeMap<String, String>,
) -> anyhow::Result<()> {
    let url = apiserver_url.join(&format!("/api/v1/nodes/{node_name}"))?;
    client
        .patch(url)
        .header("content-type", "application/merge-patch+json")
        .body(serde_json::to_vec(
            &json!({ "metadata": { "labels": labels } }),
        )?)
        .send()
        .await?
        .error_for_status()
        .context("fake apiserver patch failed")?;
    Ok(())
}

async fn get_node(
    client: &Client,
    apiserver_url: &Url,
    node_name: &str,
) -> anyhow::Result<serde_json::Value> {
    let url = apiserver_url.join(&format!("/api/v1/nodes/{node_name}"))?;
    let body = client
        .get(url)
        .send()
        .await?
        .error_for_status()?
        .text()
        .await?;
    Ok(serde_json::from_str(&body)?)
}

fn label_map(node: &serde_json::Value) -> BTreeMap<String, String> {
    node.get("metadata")
        .and_then(|metadata| metadata.get("labels"))
        .and_then(serde_json::Value::as_object)
        .map(|labels| {
            labels
                .iter()
                .filter_map(|(key, value)| {
                    value.as_str().map(|value| (key.clone(), value.to_string()))
                })
                .collect()
        })
        .unwrap_or_default()
}
